pkg/util/debouncer/debouncer_test.go

```diff
@@ -0,0 +1,214 @@
+		// Buffer should be full by now as we are not reading from it yet.
+		require.ErrorIs(t, group.Add("key2"), ErrBufferFull)
+	})
+
+	t.Run("should track metrics", func(t *testing.T) {
+		var wg sync.WaitGroup
+
+		group, err := NewGroup(DebouncerOpts[string]{
+			BufferSize: 10,
+			KeyFunc:    func(s string) string { return s },
+			ProcessHandler: func(ctx context.Context, value string) error {
+				wg.Done()
+				return nil
+			},
+			MinWait: 10 * time.Millisecond,
+			MaxWait: 100 * time.Millisecond,
+			Reg:     prometheus.NewPedanticRegistry(),
+			Name:    "test",
+		})
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		group.Start(ctx)
+
+		wg.Add(1)
+		require.NoError(t, group.Add("key1"))
+		require.NoError(t, group.Add("key1"))
+
+		wg.Wait()
+
+		require.Equal(t, float64(2), testutil.ToFloat64(group.metrics.itemsAddedCounter))
+		require.Equal(t, float64(1), testutil.ToFloat64(group.metrics.itemsProcessedCounter))
+	})
+
+	t.Run("should handle errors", func(t *testing.T) {
+		var (
+			wg          sync.WaitGroup
+			errs        = make(chan error, 10)
+			expectedErr = errors.New("test error")
+		)
+
+		group, err := NewGroup(DebouncerOpts[string]{
+			BufferSize: 10,
+			KeyFunc:    func(s string) string { return s },
+			ProcessHandler: func(ctx context.Context, value string) error {
+				wg.Done()
+				return expectedErr
+			},
+			MinWait:      10 * time.Millisecond,
+			MaxWait:      100 * time.Millisecond,
+			Reg:          prometheus.NewPedanticRegistry(),
+			Name:         "test_errors",
+			ErrorHandler: func(_ string, err error) { errs <- err },
+		})
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		group.Start(ctx)
+
+		wg.Add(1)
+		require.NoError(t, group.Add("key1"))
+
+		wg.Wait()
+
+		select {
+		case err := <-errs:
+			require.Equal(t, expectedErr, err)
+		default:
+			t.Fatal("expected error")
+		}
+
+		require.Equal(t, float64(1), testutil.ToFloat64(group.metrics.processingErrorsCounter))
+	})
+
+	t.Run("should gracefully handle stops", func(t *testing.T) {
+		// Create a channel to signal when processing is done.
+		done := make(chan struct{})
+
+		group, err := NewGroup(DebouncerOpts[string]{
+			BufferSize: 10,
+			KeyFunc:    func(s string) string { return s },
+			ProcessHandler: func(ctx context.Context, item string) error {
+				// Start a goroutine to wait for context cancellation.
+				go func() {
+					<-ctx.Done()
+					close(done)
+				}()
+				return nil
+			},
+			MinWait: 50 * time.Millisecond,
+			MaxWait: 100 * time.Millisecond,
+		})
+		require.NoError(t, err)
+
+		// Start the group with a context
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		group.Start(ctx)
+
+		// Send an item to trigger processing.
+		require.NoError(t, group.Add("key-1"))
+
+		// Give the group a moment to process the item.
+		time.Sleep(50 * time.Millisecond)
```

Comment:
This seems very close to min time. It is possible that `group.Stop` will be called before debouncer's minTime timer fires, and debouncer will react on canceled context instead, making this test flaky.